// @astrojs/language-server/dist/plugins/astro/AstroPlugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AstroPlugin = void 0;
const vscode_languageserver_1 = require("vscode-languageserver");
const utils_1 = require("../../utils");
const utils_2 = require("../typescript/utils");
const utils_3 = require("../../core/documents/utils");
const ts = __importStar(require("typescript"));
const LanguageServiceManager_1 = require("../typescript/LanguageServiceManager");
const utils_4 = require("../typescript/utils");
const vscode_languageserver_types_1 = require("vscode-languageserver-types");
const CompletionProvider_1 = require("./features/CompletionProvider");
class AstroPlugin {
    constructor(docManager, configManager, workspaceUris) {
        this.pluginName = 'Astro';
        this.configManager = configManager;
        this.tsLanguageServiceManager = new LanguageServiceManager_1.LanguageServiceManager(docManager, configManager, workspaceUris);
        this.completionProvider = new CompletionProvider_1.CompletionProvider(docManager, this.tsLanguageServiceManager);
    }
    async getCompletions(document, position, completionContext) {
        const completions = this.completionProvider.getCompletions(document, position, completionContext);
        return completions;
    }
    async getFoldingRanges(document) {
        const foldingRanges = [];
        const { frontmatter } = document.astro;
        // Currently editing frontmatter, don't fold
        if (frontmatter.state !== 'closed')
            return foldingRanges;
        const start = document.positionAt(frontmatter.startOffset);
        const end = document.positionAt(frontmatter.endOffset - 3);
        return [
            {
                startLine: start.line,
                startCharacter: start.character,
                endLine: end.line,
                endCharacter: end.character,
                kind: vscode_languageserver_types_1.FoldingRangeKind.Imports,
            },
        ];
    }
    async getDefinitions(document, position) {
        if (this.isInsideFrontmatter(document, position)) {
            return [];
        }
        const offset = document.offsetAt(position);
        const html = document.html;
        const node = html.findNodeAt(offset);
        if (!this.isComponentTag(node)) {
            return [];
        }
        const [componentName] = node.tag.split(':');
        const { lang } = await this.tsLanguageServiceManager.getTypeScriptDoc(document);
        const defs = this.getDefinitionsForComponentName(document, lang, componentName);
        if (!defs || !defs.length) {
            return [];
        }
        const startRange = vscode_languageserver_1.Range.create(vscode_languageserver_1.Position.create(0, 0), vscode_languageserver_1.Position.create(0, 0));
        const links = defs.map((def) => {
            const defFilePath = (0, utils_4.ensureRealFilePath)(def.fileName);
            return vscode_languageserver_1.LocationLink.create((0, utils_1.pathToUrl)(defFilePath), startRange, startRange);
        });
        return links;
    }
    isInsideFrontmatter(document, position) {
        return (0, utils_3.isInsideFrontmatter)(document.getText(), document.offsetAt(position));
    }
    isComponentTag(node) {
        if (!node.tag) {
            return false;
        }
        const firstChar = node.tag[0];
        return /[A-Z]/.test(firstChar);
    }
    getDefinitionsForComponentName(document, lang, componentName) {
        const filePath = (0, utils_1.urlToPath)(document.uri);
        const tsFilePath = (0, utils_2.toVirtualAstroFilePath)(filePath);
        const program = lang.getProgram();
        const sourceFile = program === null || program === void 0 ? void 0 : program.getSourceFile(tsFilePath);
        if (!sourceFile) {
            return undefined;
        }
        const specifier = this.getImportSpecifierForIdentifier(sourceFile, componentName);
        if (!specifier) {
            return [];
        }
        const defs = lang.getDefinitionAtPosition(tsFilePath, specifier.getStart());
        if (!defs) {
            return undefined;
        }
        return defs;
    }
    getImportSpecifierForIdentifier(sourceFile, identifier) {
        let importSpecifier = undefined;
        ts.forEachChild(sourceFile, (tsNode) => {
            if (ts.isImportDeclaration(tsNode)) {
                if (tsNode.importClause) {
                    const { name, namedBindings } = tsNode.importClause;
                    if (name && name.getText() === identifier) {
                        importSpecifier = tsNode.moduleSpecifier;
                        return true;
                    }
                    else if (namedBindings && namedBindings.kind === ts.SyntaxKind.NamedImports) {
                        const elements = namedBindings.elements;
                        for (let elem of elements) {
                            if (elem.name.getText() === identifier) {
                                importSpecifier = tsNode.moduleSpecifier;
                                return true;
                            }
                        }
                    }
                }
            }
        });
        return importSpecifier;
    }
}
exports.AstroPlugin = AstroPlugin;
function isNodeExported(node) {
    return (ts.getCombinedModifierFlags(node) & ts.ModifierFlags.Export) !== 0 || (!!node.parent && node.parent.kind === ts.SyntaxKind.SourceFile);
}
