// @astrojs/language-server/dist/plugins/typescript/module-loader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAstroModuleLoader = void 0;
const typescript_1 = __importDefault(require("typescript"));
const utils_1 = require("./utils");
const astro_sys_1 = require("./astro-sys");
/**
 * Caches resolved modules.
 */
class ModuleResolutionCache {
    constructor() {
        this.cache = new Map();
    }
    /**
     * Tries to get a cached module.
     */
    get(moduleName, containingFile) {
        return this.cache.get(this.getKey(moduleName, containingFile));
    }
    /**
     * Caches resolved module, if it is not undefined.
     */
    set(moduleName, containingFile, resolvedModule) {
        if (!resolvedModule) {
            return;
        }
        this.cache.set(this.getKey(moduleName, containingFile), resolvedModule);
    }
    /**
     * Deletes module from cache. Call this if a file was deleted.
     * @param resolvedModuleName full path of the module
     */
    delete(resolvedModuleName) {
        this.cache.forEach((val, key) => {
            if (val.resolvedFileName === resolvedModuleName) {
                this.cache.delete(key);
            }
        });
    }
    getKey(moduleName, containingFile) {
        return containingFile + ':::' + (0, utils_1.ensureRealAstroFilePath)(moduleName);
    }
}
/**
 * Creates a module loader specifically for `.astro` files.
 *
 * The typescript language service tries to look up other files that are referenced in the currently open astro file.
 * For `.ts`/`.js` files this works, for `.astro` files it does not by default.
 * Reason: The typescript language service does not know about the `.astro` file ending,
 * so it assumes it's a normal typescript file and searches for files like `../Component.astro.ts`, which is wrong.
 * In order to fix this, we need to wrap typescript's module resolution and reroute all `.astro.ts` file lookups to .astro.
 *
 * @param getSnapshot A function which returns a (in case of astro file fully preprocessed) typescript/javascript snapshot
 * @param compilerOptions The typescript compiler options
 */
function createAstroModuleLoader(getSnapshot, compilerOptions) {
    const astroSys = (0, astro_sys_1.createAstroSys)(getSnapshot);
    const moduleCache = new ModuleResolutionCache();
    return {
        fileExists: astroSys.fileExists,
        readFile: astroSys.readFile,
        writeFile: astroSys.writeFile,
        readDirectory: astroSys.readDirectory,
        directoryExists: astroSys.directoryExists,
        getDirectories: astroSys.getDirectories,
        realpath: astroSys.realpath,
        deleteFromModuleCache: (path) => moduleCache.delete(path),
        resolveModuleNames,
    };
    function resolveModuleNames(moduleNames, containingFile) {
        return moduleNames.map((moduleName) => {
            const cachedModule = moduleCache.get(moduleName, containingFile);
            if (cachedModule) {
                return cachedModule;
            }
            const resolvedModule = resolveModuleName(moduleName, containingFile);
            moduleCache.set(moduleName, containingFile, resolvedModule);
            return resolvedModule;
        });
    }
    function resolveModuleName(name, containingFile) {
        // Delegate to the TS resolver first.
        // If that does not bring up anything, try the Astro Module loader
        // which is able to deal with .astro files.
        const tsResolvedModule = typescript_1.default.resolveModuleName(name, containingFile, compilerOptions, typescript_1.default.sys).resolvedModule;
        if (tsResolvedModule && !(0, utils_1.isVirtualAstroFilePath)(tsResolvedModule.resolvedFileName)) {
            return tsResolvedModule;
        }
        const astroResolvedModule = typescript_1.default.resolveModuleName(name, containingFile, compilerOptions, astroSys).resolvedModule;
        if (!astroResolvedModule || !(0, utils_1.isVirtualAstroFilePath)(astroResolvedModule.resolvedFileName)) {
            return astroResolvedModule;
        }
        const resolvedFileName = (0, utils_1.ensureRealAstroFilePath)(astroResolvedModule.resolvedFileName);
        const snapshot = getSnapshot(resolvedFileName);
        const resolvedastroModule = {
            extension: (0, utils_1.getExtensionFromScriptKind)(snapshot && snapshot.scriptKind),
            resolvedFileName,
        };
        return resolvedastroModule;
    }
}
exports.createAstroModuleLoader = createAstroModuleLoader;
