// @astrojs/language-server/dist/plugins/typescript/TypeScriptPlugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeScriptPlugin = void 0;
const path_1 = require("path");
const typescript_1 = require("typescript");
const vscode_languageserver_1 = require("vscode-languageserver");
const ts = __importStar(require("typescript"));
const LanguageServiceManager_1 = require("./LanguageServiceManager");
const utils_1 = require("./utils");
const utils_2 = require("../../utils");
const CompletionsProvider_1 = require("./features/CompletionsProvider");
const HoverProvider_1 = require("./features/HoverProvider");
const DiagnosticsProvider_1 = require("./features/DiagnosticsProvider");
const utils_3 = require("./features/utils");
const SignatureHelpProvider_1 = require("./features/SignatureHelpProvider");
class TypeScriptPlugin {
    constructor(docManager, configManager, workspaceUris) {
        this.pluginName = 'TypeScript';
        this.docManager = docManager;
        this.configManager = configManager;
        this.languageServiceManager = new LanguageServiceManager_1.LanguageServiceManager(docManager, configManager, workspaceUris);
        this.completionProvider = new CompletionsProvider_1.CompletionsProviderImpl(this.languageServiceManager);
        this.hoverProvider = new HoverProvider_1.HoverProviderImpl(this.languageServiceManager);
        this.signatureHelpProvider = new SignatureHelpProvider_1.SignatureHelpProviderImpl(this.languageServiceManager);
        this.diagnosticsProvider = new DiagnosticsProvider_1.DiagnosticsProviderImpl(this.languageServiceManager);
    }
    async doHover(document, position) {
        return this.hoverProvider.doHover(document, position);
    }
    async rename(document, position, newName) {
        const { lang, tsDoc } = await this.languageServiceManager.getTypeScriptDoc(document);
        const fragment = await tsDoc.getFragment();
        const offset = fragment.offsetAt(fragment.getGeneratedPosition(position));
        let renames = lang.findRenameLocations((0, utils_1.toVirtualAstroFilePath)(tsDoc.filePath), offset, false, false, true);
        if (!renames) {
            return null;
        }
        let edit = {
            changes: {}
        };
        renames.forEach(rename => {
            const filePath = (0, utils_1.ensureRealAstroFilePath)(rename.fileName);
            if (!(filePath in edit.changes)) {
                edit.changes[filePath] = [];
            }
            edit.changes[filePath].push({
                newText: newName,
                range: (0, utils_1.convertToLocationRange)(fragment, rename.textSpan)
            });
        });
        return edit;
    }
    async getCompletions(document, position, completionContext) {
        const completions = await this.completionProvider.getCompletions(document, position, completionContext);
        return completions;
    }
    async resolveCompletion(document, completionItem) {
        return this.completionProvider.resolveCompletion(document, completionItem);
    }
    async getDefinitions(document, position) {
        const { lang, tsDoc } = await this.languageServiceManager.getTypeScriptDoc(document);
        const mainFragment = await tsDoc.getFragment();
        const filePath = tsDoc.filePath;
        const tsFilePath = (0, utils_1.toVirtualAstroFilePath)(filePath);
        const fragmentPosition = mainFragment.getGeneratedPosition(position);
        const fragmentOffset = mainFragment.offsetAt(fragmentPosition);
        let defs = lang.getDefinitionAndBoundSpan(tsFilePath, fragmentOffset);
        if (!defs || !defs.definitions) {
            return [];
        }
        // Resolve all imports if we can
        if (this.goToDefinitionFoundOnlyAlias(tsFilePath, defs.definitions)) {
            let importDef = this.getGoToDefinitionRefsForImportSpecifier(tsFilePath, fragmentOffset, lang);
            if (importDef) {
                defs = importDef;
            }
        }
        const docs = new utils_3.SnapshotFragmentMap(this.languageServiceManager);
        docs.set(tsDoc.filePath, { fragment: mainFragment, snapshot: tsDoc });
        const result = await Promise.all(defs.definitions.map(async (def) => {
            const { fragment, snapshot } = await docs.retrieve(def.fileName);
            if ((0, utils_3.isNoTextSpanInGeneratedCode)(snapshot.getFullText(), def.textSpan)) {
                const fileName = (0, utils_1.ensureRealAstroFilePath)(def.fileName);
                const textSpan = (0, utils_1.isVirtualAstroFilePath)(tsFilePath) ? { start: 0, length: 0 } : def.textSpan;
                return vscode_languageserver_1.LocationLink.create((0, utils_2.pathToUrl)(fileName), (0, utils_1.convertToLocationRange)(fragment, textSpan), (0, utils_1.convertToLocationRange)(fragment, textSpan), (0, utils_1.convertToLocationRange)(mainFragment, defs.textSpan));
            }
        }));
        return result.filter(utils_2.isNotNullOrUndefined);
    }
    async getDiagnostics(document, cancellationToken) {
        if (!this.featureEnabled('diagnostics')) {
            return [];
        }
        return this.diagnosticsProvider.getDiagnostics(document, cancellationToken);
    }
    async onWatchFileChanges(onWatchFileChangesParams) {
        const doneUpdateProjectFiles = new Set();
        for (const { fileName, changeType } of onWatchFileChangesParams) {
            const scriptKind = (0, utils_1.getScriptKindFromFileName)(fileName);
            if (scriptKind === ts.ScriptKind.Unknown) {
                // We don't deal with svelte files here
                continue;
            }
            const snapshotManager = await this.getSnapshotManager(fileName);
            if (changeType === vscode_languageserver_1.FileChangeType.Created) {
                if (!doneUpdateProjectFiles.has(snapshotManager)) {
                    snapshotManager.updateProjectFiles();
                    doneUpdateProjectFiles.add(snapshotManager);
                }
            }
            else if (changeType === vscode_languageserver_1.FileChangeType.Deleted) {
                snapshotManager.delete(fileName);
                return;
            }
            snapshotManager.updateProjectFile(fileName);
        }
    }
    async getSignatureHelp(document, position, context, cancellationToken) {
        return this.signatureHelpProvider.getSignatureHelp(document, position, context, cancellationToken);
    }
    /**
     *
     * @internal
     */
    async getSnapshotManager(fileName) {
        return this.languageServiceManager.getSnapshotManager(fileName);
    }
    goToDefinitionFoundOnlyAlias(tsFileName, defs) {
        return !!(defs.length === 1 && defs[0].kind === 'alias' && defs[0].fileName === tsFileName);
    }
    getGoToDefinitionRefsForImportSpecifier(tsFilePath, offset, lang) {
        const program = lang.getProgram();
        const sourceFile = program === null || program === void 0 ? void 0 : program.getSourceFile(tsFilePath);
        if (sourceFile) {
            let node = ts.getTouchingPropertyName(sourceFile, offset);
            if (node && node.kind === typescript_1.SyntaxKind.Identifier) {
                if (node.parent.kind === typescript_1.SyntaxKind.ImportClause) {
                    let decl = node.parent.parent;
                    let spec = ts.isStringLiteral(decl.moduleSpecifier) && decl.moduleSpecifier.text;
                    if (spec) {
                        let fileName = (0, path_1.join)((0, path_1.dirname)(tsFilePath), spec);
                        let start = node.pos + 1;
                        let def = {
                            definitions: [
                                {
                                    kind: 'alias',
                                    fileName,
                                    name: '',
                                    containerKind: '',
                                    containerName: '',
                                    textSpan: {
                                        start: 0,
                                        length: 0,
                                    },
                                },
                            ],
                            textSpan: {
                                start,
                                length: node.end - start,
                            },
                        };
                        return def;
                    }
                }
            }
        }
    }
    // This exists so we can make features toggleable in the future.
    featureEnabled(feature) {
        return (this.configManager.enabled('typescript.enable') &&
            this.configManager.enabled(`typescript.${feature}.enable`));
    }
}
exports.TypeScriptPlugin = TypeScriptPlugin;
