// @astrojs/language-server/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.debounceThrottle = exports.debounceSameArg = exports.isBeforeOrEqualToPosition = exports.isInRange = exports.isNotNullOrUndefined = exports.clamp = exports.flatten = exports.isPossibleClientComponent = exports.isPossibleComponent = exports.pathToUrl = exports.urlToPath = exports.normalizeUri = void 0;
const vscode_uri_1 = require("vscode-uri");
/** Normalizes a document URI */
function normalizeUri(uri) {
    return vscode_uri_1.URI.parse(uri).toString();
}
exports.normalizeUri = normalizeUri;
/** Turns a URL into a normalized FS Path */
function urlToPath(stringUrl) {
    const url = vscode_uri_1.URI.parse(stringUrl);
    if (url.scheme !== 'file') {
        return null;
    }
    return url.fsPath.replace(/\\/g, '/');
}
exports.urlToPath = urlToPath;
/** Converts a path to a URL */
function pathToUrl(path) {
    return vscode_uri_1.URI.file(path).toString();
}
exports.pathToUrl = pathToUrl;
/**
 *
 * The language service is case insensitive, and would provide
 * hover info for Svelte components like `Option` which have
 * the same name like a html tag.
 */
function isPossibleComponent(node) {
    var _a;
    return !!((_a = node.tag) === null || _a === void 0 ? void 0 : _a[0].match(/[A-Z]/));
}
exports.isPossibleComponent = isPossibleComponent;
/**
 *
 * The language service is case insensitive, and would provide
 * hover info for Svelte components like `Option` which have
 * the same name like a html tag.
 */
function isPossibleClientComponent(node) {
    if (isPossibleComponent(node) && node.attributes) {
        for (let [name] of Object.entries(node.attributes)) {
            if (name.startsWith('client:')) {
                return true;
            }
        }
    }
    return false;
}
exports.isPossibleClientComponent = isPossibleClientComponent;
/** Flattens an array */
function flatten(arr) {
    return arr.reduce((all, item) => [...all, ...item], []);
}
exports.flatten = flatten;
/** Clamps a number between min and max */
function clamp(num, min, max) {
    return Math.max(min, Math.min(max, num));
}
exports.clamp = clamp;
function isNotNullOrUndefined(val) {
    return val !== undefined && val !== null;
}
exports.isNotNullOrUndefined = isNotNullOrUndefined;
/** Checks if a position is inside range */
function isInRange(positionToTest, range) {
    return isBeforeOrEqualToPosition(range.end, positionToTest) && isBeforeOrEqualToPosition(positionToTest, range.start);
}
exports.isInRange = isInRange;
/**  */
function isBeforeOrEqualToPosition(position, positionToTest) {
    return positionToTest.line < position.line || (positionToTest.line === position.line && positionToTest.character <= position.character);
}
exports.isBeforeOrEqualToPosition = isBeforeOrEqualToPosition;
/**
 * Debounces a function but cancels previous invocation only if
 * a second function determines it should.
 *
 * @param fn The function with it's argument
 * @param determineIfSame The function which determines if the previous invocation should be canceld or not
 * @param miliseconds Number of miliseconds to debounce
 */
function debounceSameArg(fn, shouldCancelPrevious, miliseconds) {
    let timeout;
    let prevArg;
    return (arg) => {
        if (shouldCancelPrevious(arg, prevArg)) {
            clearTimeout(timeout);
        }
        prevArg = arg;
        timeout = setTimeout(() => {
            fn(arg);
            prevArg = undefined;
        }, miliseconds);
    };
}
exports.debounceSameArg = debounceSameArg;
/**
 * Debounces a function but also waits at minimum the specified number of miliseconds until
 * the next invocation. This avoids needless calls when a synchronous call (like diagnostics)
 * took too long and the whole timeout of the next call was eaten up already.
 *
 * @param fn The function with it's argument
 * @param miliseconds Number of miliseconds to debounce/throttle
 */
function debounceThrottle(fn, miliseconds) {
    let timeout;
    let lastInvocation = Date.now() - miliseconds;
    function maybeCall(...args) {
        clearTimeout(timeout);
        timeout = setTimeout(() => {
            if (Date.now() - lastInvocation < miliseconds) {
                maybeCall(...args);
                return;
            }
            fn(...args);
            lastInvocation = Date.now();
        }, miliseconds);
    }
    return maybeCall;
}
exports.debounceThrottle = debounceThrottle;
