// astro/dist/vite-plugin-astro/index.js
import esbuild from "esbuild";
import fs from "fs";
import { fileURLToPath } from "url";
import { transform } from "@astrojs/compiler";
import { getViteTransform, transformWithVite } from "./styles.js";
const FRONTMATTER_PARSE_REGEXP = /^\-\-\-(.*)^\-\-\-/ms;
function isSSR(options) {
  if (options === void 0) {
    return false;
  }
  if (typeof options === "boolean") {
    return options;
  }
  if (typeof options == "object") {
    return !!options.ssr;
  }
  return false;
}
function astro({ config, devServer }) {
  let viteTransform;
  return {
    name: "@astrojs/vite-plugin-astro",
    enforce: "pre",
    configResolved(resolvedConfig) {
      viteTransform = getViteTransform(resolvedConfig);
    },
    async load(id, opts) {
      if (!id.endsWith(".astro")) {
        return null;
      }
      const normalizedID = fileURLToPath(new URL(`file://${id}`));
      const isPage = normalizedID.startsWith(fileURLToPath(config.pages)) || normalizedID.startsWith(fileURLToPath(config.layouts));
      let source = await fs.promises.readFile(id, "utf8");
      let tsResult;
      let cssTransformError;
      try {
        tsResult = await transform(source, {
          as: isPage ? "document" : "fragment",
          projectRoot: config.projectRoot.toString(),
          site: config.buildOptions.site,
          sourcefile: id,
          sourcemap: "both",
          internalURL: "astro/internal",
          preprocessStyle: async (value, attrs) => {
            const lang = `.${(attrs == null ? void 0 : attrs.lang) || "css"}`.toLowerCase();
            try {
              const result = await transformWithVite({ value, lang, id, transformHook: viteTransform, ssr: isSSR(opts) });
              let map2;
              if (!result)
                return null;
              if (result.map) {
                if (typeof result.map === "string") {
                  map2 = result.map;
                } else if (result.map.mappings) {
                  map2 = result.map.toString();
                }
              }
              return { code: result.code, map: map2 };
            } catch (err) {
              cssTransformError = err;
              return null;
            }
          }
        });
        if (cssTransformError)
          throw cssTransformError;
        const { code, map } = await esbuild.transform(tsResult.code, { loader: "ts", sourcemap: "external", sourcefile: id });
        return {
          code,
          map
        };
      } catch (err) {
        const scannedFrontmatter = FRONTMATTER_PARSE_REGEXP.exec(source);
        if (scannedFrontmatter) {
          try {
            await esbuild.transform(scannedFrontmatter[1], { loader: "ts", sourcemap: false, sourcefile: id });
          } catch (frontmatterErr) {
            if (frontmatterErr && frontmatterErr.message) {
              frontmatterErr.message = frontmatterErr.message.replace("end of file", "end of frontmatter");
            }
            throw frontmatterErr;
          }
        }
        if (err.stack.includes("wasm-function")) {
          const search = new URLSearchParams({
            labels: "compiler",
            title: "\u{1F41B} BUG: `@astrojs/compiler` panic",
            body: `### Describe the Bug

\`@astrojs/compiler\` encountered an unrecoverable error when compiling the following file.

**${id.replace(fileURLToPath(config.projectRoot), "")}**
\`\`\`astro
${source}
\`\`\`
`
          });
          err.url = `https://github.com/withastro/astro/issues/new?${search.toString()}`;
          err.message = `Error: Uh oh, the Astro compiler encountered an unrecoverable error!

Please open
a GitHub issue using the link below:
${err.url}`;
          err.stack = `    at ${id}`;
        }
        throw err;
      }
    }
  };
}
export {
  astro as default
};
