// @astrojs/language-server/dist/core/documents/DocumentManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentManager = void 0;
const events_1 = require("events");
const Document_1 = require("./Document");
const utils_1 = require("../../utils");
class DocumentManager {
    constructor(createDocument) {
        this.createDocument = createDocument;
        this.emitter = new events_1.EventEmitter();
        this.openedInClient = new Set();
        this.documents = new Map();
        this.locked = new Set();
        this.deleteCandidates = new Set();
    }
    get(uri) {
        return this.documents.get((0, utils_1.normalizeUri)(uri));
    }
    openDocument(textDocument) {
        let document;
        if (this.documents.has(textDocument.uri)) {
            document = this.get(textDocument.uri);
            if (textDocument.overrideText)
                document.setText(textDocument.text);
        }
        else {
            document = this.createDocument(textDocument);
            this.documents.set((0, utils_1.normalizeUri)(textDocument.uri), document);
            this.notify('documentOpen', document);
        }
        this.notify('documentChange', document);
        return document;
    }
    closeDocument(uri) {
        uri = (0, utils_1.normalizeUri)(uri);
        const document = this.documents.get(uri);
        if (!document) {
            throw new Error('Cannot call methods on an unopened document');
        }
        this.notify('documentClose', document);
        // Some plugin may prevent a document from actually being closed.
        if (!this.locked.has(uri)) {
            this.documents.delete(uri);
        }
        else {
            this.deleteCandidates.add(uri);
        }
        this.openedInClient.delete(uri);
    }
    releaseDocument(uri) {
        uri = (0, utils_1.normalizeUri)(uri);
        this.locked.delete(uri);
        this.openedInClient.delete(uri);
        if (this.deleteCandidates.has(uri)) {
            this.deleteCandidates.delete(uri);
            this.closeDocument(uri);
        }
    }
    updateDocument(uri, changes) {
        const document = this.documents.get((0, utils_1.normalizeUri)(uri));
        if (!document) {
            throw new Error('Cannot call methods on an unopened document');
        }
        for (const change of changes) {
            let start = 0;
            let end = 0;
            if ('range' in change) {
                start = document.offsetAt(change.range.start);
                end = document.offsetAt(change.range.end);
            }
            else {
                end = document.getTextLength();
            }
            document.update(change.text, start, end);
        }
        this.notify('documentChange', document);
    }
    markAsOpenedInClient(uri) {
        this.openedInClient.add((0, utils_1.normalizeUri)(uri));
    }
    getAllOpenedByClient() {
        return Array.from(this.documents.entries()).filter((doc) => this.openedInClient.has(doc[0]));
    }
    on(name, listener) {
        this.emitter.on(name, listener);
    }
    notify(name, document) {
        this.emitter.emit(name, document);
    }
    static newInstance() {
        return new DocumentManager(({ uri, text }) => new Document_1.Document(uri, text));
    }
}
exports.DocumentManager = DocumentManager;
