// @astrojs/language-server/dist/core/documents/Document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Document = void 0;
const vscode_languageserver_1 = require("vscode-languageserver");
const utils_1 = require("../../utils");
const parseHtml_1 = require("./parseHtml");
const parseAstro_1 = require("./parseAstro");
const utils_2 = require("./utils");
class Document {
    constructor(uri, text) {
        this.uri = uri;
        this.languageId = 'astro';
        this.version = 0;
        this.styleInfo = null;
        this.content = text;
        this.updateDocInfo();
    }
    updateDocInfo() {
        this.html = (0, parseHtml_1.parseHtml)(this.content);
        this.astro = (0, parseAstro_1.parseAstro)(this.content);
        this.styleInfo = (0, utils_2.extractStyleTag)(this.content, this.html);
        if (this.styleInfo) {
            this.styleInfo.attributes.lang = 'css';
        }
    }
    setText(text) {
        this.content = text;
        this.version++;
        this.updateDocInfo();
    }
    /**
     * Update the text between two positions.
     * @param text The new text slice
     * @param start Start offset of the new text
     * @param end End offset of the new text
     */
    update(text, start, end) {
        const content = this.getText();
        this.setText(content.slice(0, start) + text + content.slice(end));
    }
    getText() {
        return this.content;
    }
    /**
     * Get the line and character based on the offset
     * @param offset The index of the position
     */
    positionAt(offset) {
        offset = (0, utils_1.clamp)(offset, 0, this.getTextLength());
        const lineOffsets = this.getLineOffsets();
        let low = 0;
        let high = lineOffsets.length;
        if (high === 0) {
            return vscode_languageserver_1.Position.create(0, offset);
        }
        while (low < high) {
            const mid = Math.floor((low + high) / 2);
            if (lineOffsets[mid] > offset) {
                high = mid;
            }
            else {
                low = mid + 1;
            }
        }
        // low is the least x for which the line offset is larger than the current offset
        // or array.length if no line offset is larger than the current offset
        const line = low - 1;
        return vscode_languageserver_1.Position.create(line, offset - lineOffsets[line]);
    }
    /**
     * Get the index of the line and character position
     * @param position Line and character position
     */
    offsetAt(position) {
        const lineOffsets = this.getLineOffsets();
        if (position.line >= lineOffsets.length) {
            return this.getTextLength();
        }
        else if (position.line < 0) {
            return 0;
        }
        const lineOffset = lineOffsets[position.line];
        const nextLineOffset = position.line + 1 < lineOffsets.length ? lineOffsets[position.line + 1] : this.getTextLength();
        return (0, utils_1.clamp)(nextLineOffset, lineOffset, lineOffset + position.character);
    }
    getLineUntilOffset(offset) {
        const { line, character } = this.positionAt(offset);
        return this.lines[line].slice(0, character);
    }
    getLineOffsets() {
        const lineOffsets = [];
        const text = this.getText();
        let isLineStart = true;
        for (let i = 0; i < text.length; i++) {
            if (isLineStart) {
                lineOffsets.push(i);
                isLineStart = false;
            }
            const ch = text.charAt(i);
            isLineStart = ch === '\r' || ch === '\n';
            if (ch === '\r' && i + 1 < text.length && text.charAt(i + 1) === '\n') {
                i++;
            }
        }
        if (isLineStart && text.length > 0) {
            lineOffsets.push(text.length);
        }
        return lineOffsets;
    }
    /**
     * Get the length of the document's content
     */
    getTextLength() {
        return this.getText().length;
    }
    /**
     * Returns the file path if the url scheme is file
     */
    getFilePath() {
        return (0, utils_1.urlToPath)(this.uri);
    }
    /**
     * Get URL file path.
     */
    getURL() {
        return this.uri;
    }
    get lines() {
        return this.getText().split(/\r?\n/);
    }
    get lineCount() {
        return this.lines.length;
    }
}
exports.Document = Document;
