// @astrojs/language-server/dist/core/documents/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractStyleTag = exports.walk = exports.offsetAt = exports.positionAt = exports.isInTag = exports.isInsideFrontmatter = exports.isInsideExpression = exports.getFirstNonWhitespaceIndex = exports.getWordAt = exports.getWordRangeAt = void 0;
const vscode_html_languageservice_1 = require("vscode-html-languageservice");
const vscode_languageserver_1 = require("vscode-languageserver");
const utils_1 = require("../../utils");
const parseHtml_1 = require("./parseHtml");
function parseAttributes(rawAttrs) {
    const attrs = {};
    if (!rawAttrs) {
        return attrs;
    }
    Object.keys(rawAttrs).forEach((attrName) => {
        const attrValue = rawAttrs[attrName];
        attrs[attrName] = attrValue === null ? attrName : removeOuterQuotes(attrValue);
    });
    return attrs;
    function removeOuterQuotes(attrValue) {
        if ((attrValue.startsWith('"') && attrValue.endsWith('"')) || (attrValue.startsWith("'") && attrValue.endsWith("'"))) {
            return attrValue.slice(1, attrValue.length - 1);
        }
        return attrValue;
    }
}
/**
 * Gets word range at position.
 * Delimiter is by default a whitespace, but can be adjusted.
 */
function getWordRangeAt(str, pos, delimiterRegex = { left: /\S+$/, right: /\s/ }) {
    let start = str.slice(0, pos).search(delimiterRegex.left);
    if (start < 0) {
        start = pos;
    }
    let end = str.slice(pos).search(delimiterRegex.right);
    if (end < 0) {
        end = str.length;
    }
    else {
        end = end + pos;
    }
    return { start, end };
}
exports.getWordRangeAt = getWordRangeAt;
/**
 * Gets word at position.
 * Delimiter is by default a whitespace, but can be adjusted.
 */
function getWordAt(str, pos, delimiterRegex = { left: /\S+$/, right: /\s/ }) {
    const { start, end } = getWordRangeAt(str, pos, delimiterRegex);
    return str.slice(start, end);
}
exports.getWordAt = getWordAt;
/**
 * Gets index of first-non-whitespace character.
 */
function getFirstNonWhitespaceIndex(str) {
    return str.length - str.trimStart().length;
}
exports.getFirstNonWhitespaceIndex = getFirstNonWhitespaceIndex;
/** checks if a position is currently inside of an expression */
function isInsideExpression(html, tagStart, position) {
    const charactersInNode = html.substring(tagStart, position);
    return charactersInNode.lastIndexOf('{') > charactersInNode.lastIndexOf('}');
}
exports.isInsideExpression = isInsideExpression;
/**
 * Returns if a given offset is inside of the document frontmatter
 */
function isInsideFrontmatter(text, offset) {
    let start = text.slice(0, offset).trim().split('---').length;
    let end = text.slice(offset).trim().split('---').length;
    return start > 1 && start < 3 && end >= 1;
}
exports.isInsideFrontmatter = isInsideFrontmatter;
function isInTag(position, tagInfo) {
    return !!tagInfo && (0, utils_1.isInRange)(position, vscode_languageserver_1.Range.create(tagInfo.startPos, tagInfo.endPos));
}
exports.isInTag = isInTag;
/**
 * Get the line and character based on the offset
 * @param offset The index of the position
 * @param text The text for which the position should be retrived
 */
function positionAt(offset, text) {
    offset = (0, utils_1.clamp)(offset, 0, text.length);
    const lineOffsets = getLineOffsets(text);
    let low = 0;
    let high = lineOffsets.length;
    if (high === 0) {
        return vscode_html_languageservice_1.Position.create(0, offset);
    }
    while (low < high) {
        const mid = Math.floor((low + high) / 2);
        if (lineOffsets[mid] > offset) {
            high = mid;
        }
        else {
            low = mid + 1;
        }
    }
    // low is the least x for which the line offset is larger than the current offset
    // or array.length if no line offset is larger than the current offset
    const line = low - 1;
    return vscode_html_languageservice_1.Position.create(line, offset - lineOffsets[line]);
}
exports.positionAt = positionAt;
/**
 * Get the offset of the line and character position
 * @param position Line and character position
 * @param text The text for which the offset should be retrived
 */
function offsetAt(position, text) {
    const lineOffsets = getLineOffsets(text);
    if (position.line >= lineOffsets.length) {
        return text.length;
    }
    else if (position.line < 0) {
        return 0;
    }
    const lineOffset = lineOffsets[position.line];
    const nextLineOffset = position.line + 1 < lineOffsets.length ? lineOffsets[position.line + 1] : text.length;
    return (0, utils_1.clamp)(nextLineOffset, lineOffset, lineOffset + position.character);
}
exports.offsetAt = offsetAt;
function getLineOffsets(text) {
    const lineOffsets = [];
    let isLineStart = true;
    for (let i = 0; i < text.length; i++) {
        if (isLineStart) {
            lineOffsets.push(i);
            isLineStart = false;
        }
        const ch = text.charAt(i);
        isLineStart = ch === '\r' || ch === '\n';
        if (ch === '\r' && i + 1 < text.length && text.charAt(i + 1) === '\n') {
            i++;
        }
    }
    if (isLineStart && text.length > 0) {
        lineOffsets.push(text.length);
    }
    return lineOffsets;
}
function* walk(node) {
    for (let child of node.children) {
        yield* walk(child);
    }
    yield node;
}
exports.walk = walk;
/*
export function* walk(node: Node, startIndex = 0) {
    let skip, tmp;
    let depth = 0;
    let index = startIndex;

    // Always start with the initial element.
    do {
        if ( !skip && (tmp = node.firstChild) ) {
            depth++;
            callback('child', node, tmp, index);
            index++;
        } else if ( tmp = node.nextSibling ) {
            skip = false;
            callback('sibling', node, tmp, index);
            index++;
        } else {
            tmp = node.parentNode;
            depth--;
            skip = true;
        }
        node = tmp;
    } while ( depth > 0 );
};
*/
/**
 * Extracts a tag (style or script) from the given text
 * and returns its start, end and the attributes on that tag.
 *
 * @param source text content to extract tag from
 * @param tag the tag to extract
 */
function extractTags(text, tag, html) {
    const rootNodes = (html === null || html === void 0 ? void 0 : html.roots) || (0, parseHtml_1.parseHtml)(text).roots;
    const matchedNodes = rootNodes.filter((node) => node.tag === tag);
    if (tag === 'style' && !matchedNodes.length && rootNodes.length && rootNodes[0].tag === 'html') {
        for (let child of walk(rootNodes[0])) {
            if (child.tag === 'style') {
                matchedNodes.push(child);
            }
        }
    }
    return matchedNodes.map(transformToTagInfo);
    function transformToTagInfo(matchedNode) {
        var _a, _b;
        const start = (_a = matchedNode.startTagEnd) !== null && _a !== void 0 ? _a : matchedNode.start;
        const end = (_b = matchedNode.endTagStart) !== null && _b !== void 0 ? _b : matchedNode.end;
        const startPos = positionAt(start, text);
        const endPos = positionAt(end, text);
        const container = {
            start: matchedNode.start,
            end: matchedNode.end,
        };
        const content = text.substring(start, end);
        return {
            content,
            attributes: parseAttributes(matchedNode.attributes),
            start,
            end,
            startPos,
            endPos,
            container,
        };
    }
}
function extractStyleTag(source, html) {
    const styles = extractTags(source, 'style', html);
    if (!styles.length) {
        return null;
    }
    // There can only be one style tag
    return styles[0];
}
exports.extractStyleTag = extractStyleTag;
